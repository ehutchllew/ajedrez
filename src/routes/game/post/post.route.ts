import { Request, Response } from "express";
import { createGame } from "src/services/game/post/post.service";

export function post(req: Request, res: Response) {
  try {
    const newGame = createGame();
    res.status(200).send(newGame);
  } catch (e) {
    console.error("wut wut", e);
  }
}
