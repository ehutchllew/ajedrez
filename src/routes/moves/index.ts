import { Router } from "express";

export function movesRoutes(router: Router): Router {
    return router
}