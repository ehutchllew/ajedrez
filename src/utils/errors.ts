import { HTTP_STATUS_CODE } from "./http";

abstract class BaseError {
  public readonly message: string;
  public readonly stack: string | undefined;
  public abstract statusCode: HTTP_STATUS_CODE;
  constructor(error: string | Error) {
    if (error instanceof Error) {
      this.message = error.message;
      this.stack = error.stack;
    } else {
      const err = new Error(error);
      this.message = err.message;
      this.stack = err.stack;
    }
  }
}

export class BadRequestError extends BaseError {
  public statusCode = HTTP_STATUS_CODE.BAD_REQUEST;
  constructor(error: string | Error) {
    super(error);
  }
}

export class InvalidDataError extends BaseError {
  public statusCode = HTTP_STATUS_CODE.INVALID_DATA;
  constructor(error: string | Error) {
    super(error);
  }
}

export class InternalServerError extends BaseError {
  public statusCode = HTTP_STATUS_CODE.INTERNAL_SERVER_ERROR;
  constructor(error: string | Error) {
    super(error);
  }
}
