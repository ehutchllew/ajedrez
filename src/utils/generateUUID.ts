import crypto from "node:crypto";

export function generateUUID(): string {
  return crypto.randomUUID({ disableEntropyCache: true });
}
