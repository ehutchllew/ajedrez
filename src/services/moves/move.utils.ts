import { TileNumberingType } from "src/models/board.model";
import { IGame } from "src/models/game.model";
import { IPawnMove, MoveType } from "src/models/moves.model";

/**
 *
 * TODO: Expand this to include indices of ENTIRE path. Perhaps as Array<[number, number]>
 * that way we can support pieces that aren't of type `IPawn`.
 */
export function determineIndicesOfPieceLocation(
  tile: TileNumberingType
): [number, number] {
  const [columnPosition, rowPosition] = tile.split("");
  const rowIndex = parseInt(rowPosition) - 1;
  const columnIndex = columnPosition.charCodeAt(0) - "a".charCodeAt(0);
  return [rowIndex, columnIndex];
}

/**
 *
 * FIXME: could probably get rid of this by just comparing:
 * `determineIndicesOfPieceMovement(from)` and `determineIndicesOfPieceMovement(to)`
 */
export function determineVector(
  from: IPawnMove["from"],
  to: IPawnMove["to"]
): [number, number] {
  const [fromColumn, fromRow] = from.split("");
  const [toColumn, toRow] = to.split("");

  const rowDistance = toColumn.charCodeAt(0) - fromColumn.charCodeAt(0);
  const columnDistance = parseInt(toRow) - parseInt(fromRow);

  return [rowDistance, columnDistance];
}

export function getMovedPiece(move: MoveType, gameState: IGame) {
  const [movedPieceRow, movedPieceColumn] = determineIndicesOfPieceLocation(
    move.from
  );
  const movedTileState = gameState.board[movedPieceRow][movedPieceColumn];
  return movedTileState[1];
}
