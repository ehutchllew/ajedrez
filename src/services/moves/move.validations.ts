import { TileNumberingType } from "src/models/board.model";
import { IGame } from "src/models/game.model";
import { IBishopMove, IPawnMove, MoveType } from "src/models/moves.model";
import { Empty, Pawn, PIECE_TYPE } from "src/models/piece.model";
import {
  BadRequestError,
  InternalServerError,
  InvalidDataError,
} from "src/utils/errors";

export function validateMove(move: MoveType, gameState: IGame): void {
  if (move.color !== gameState.playerTurn) {
    throw new BadRequestError(
      `Invalid move: it's ${gameState.playerTurn}'s turn, but player attempted to move a ${move.color} piece`
    );
  }

  switch (move.type) {
    case PIECE_TYPE.BISHOP:
      validateBishopMove(move);
      break;
    case PIECE_TYPE.PAWN:
      validatePawnMove(move, gameState.board);
      break;
    default:
      throw new BadRequestError(
        `Unsupported piece type. Expecting one of: (${Object.keys(PIECE_TYPE)})`
      );
  }
}

function validateBishopMove(_: IBishopMove): void {
  throw new InvalidDataError("Bishop piece not yet supported");
}

function validatePawnMove(
  move: IPawnMove,
  gameState: IGame["board"]
): IGame["board"] {
  const { color, from, to } = move;
  const [fromXIndex, fromYIndex] = determineIndicesOfPieceMovement(from);
  const fromTileState = gameState[fromXIndex][fromYIndex];
  const piece = fromTileState[1];

  if (move.type !== piece.type) {
    throw new InternalServerError(
      `We messed up somewhere:( ${move.type} ) treated as ( ${piece.type} )`
    );
  }

  const [dx, dy] = determineVector(from, to);
  if ((dy < 0 && color === "white") || (dy > 0 && color === "black")) {
    throw new BadRequestError("Invalid move: Pawn cannot move backwards.");
  }

  if (dx > dy) {
    throw new BadRequestError("Invalid move: Pawn cannot move horizontally.");
  }

  if ((dy > 1 && piece.hasMoved) || (dy > 1 && dx > 0)) {
    throw new BadRequestError(
      "Invalid move: Pawn can only move forward 2 spaces as its first move of the game."
    );
  }

  const [toXIndex, toYIndex] = determineIndicesOfPieceMovement(to);
  const toTileState = gameState[toXIndex][toYIndex];
  if (dx > 0 && toTileState[1].type === Empty().type) {
    throw new BadRequestError(
      "Invalid move: Pawn cannot move diagonally unless capturing."
    );
  }

  if (dy > 0) {
    // Temporary until `determineIndicesOfPieceMovement` supports more pieces
    const pathYIndex = move.color === "black" ? toYIndex + 1 : toYIndex - 1;
    const pathTileState = gameState[toXIndex][pathYIndex];

    if (
      toTileState[1].type !== Empty().type ||
      pathTileState[1].type !== Empty().type
    )
      throw new BadRequestError(
        "Invalid move: Pawn cannot move forward when tile is occupied by another piece."
      );
  }

  const newFromTileState = [fromTileState[0], Empty()] as typeof fromTileState;

  const newToTileState = [
    toTileState[0],
    Pawn({ color: move.color, hasMoved: true }),
  ] as typeof toTileState;

  gameState[fromXIndex][fromYIndex] = newFromTileState;
  gameState[toXIndex][toYIndex] = newToTileState;

  return gameState;
}

/**
 *
 * TODO: Expand this to include indices of ENTIRE path. Perhaps as Array<[number, number]>
 * that way we can support pieces that aren't of type `IPawn`.
 */
function determineIndicesOfPieceMovement(
  to: TileNumberingType
): [number, number] {
  const [columnPosition, rowPosition] = to.split("");
  const rowIndex = parseInt(rowPosition) - 1;
  const columnIndex = columnPosition.charCodeAt(0) - "a".charCodeAt(0);
  return [rowIndex, columnIndex];
}

/**
 *
 * FIXME: could probably get rid of this by just comparing:
 * `determineIndicesOfPieceMovement(from)` and `determineIndicesOfPieceMovement(to)`
 */
function determineVector(
  from: IPawnMove["from"],
  to: IPawnMove["to"]
): [number, number] {
  const [fromColumn, fromRow] = from.split("");
  const [toColumn, toRow] = to.split("");

  const columnDistance = toColumn.charCodeAt(0) - fromColumn.charCodeAt(0);
  const rowDistance = parseInt(toRow) - parseInt(fromRow);

  return [rowDistance, columnDistance];
}
