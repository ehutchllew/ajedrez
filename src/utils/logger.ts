import { ILogsRepository } from "src/data/logs.repository";
import { ILog } from "src/models/log.model";
import { InternalServerError } from "./errors";

export interface ILoggerParams
  extends Omit<ILog, "_id" | "createdAt" | "data"> {
  id: string;
  data: string | Record<string, any>;
}

export function createLogger(repo: ILogsRepository) {
  return {
    async createLog({
      id: _id,
      data: rawData = {},
      error: rawError = "",
      hasError,
      ips,
      method,
      params,
      path,
      query,
      statusCode,
    }: ILoggerParams) {
      try {
        const data =
          typeof rawData === "string" ? rawData : JSON.stringify(rawData);
        const error =
          typeof rawError === "string"
            ? rawError
            : JSON.stringify({
                message: rawError.message,
                name: rawError.name,
                stack: rawError.stack,
              });
        repo.createLog({
          _id,
          createdAt: Date.now(),
          data,
          error,
          hasError,
          ips,
          method,
          params,
          path,
          query,
          statusCode,
        });
      } catch (e) {
        throw new InternalServerError("unable to stringify data for logging");
      }
    },

    async getLog(_: string) {
      try {
      } finally {
      }
    },
  };
}
