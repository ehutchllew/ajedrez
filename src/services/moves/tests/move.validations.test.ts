import { MoveType } from "src/models/moves.model";
import { IEmpty, PIECE_TYPE } from "src/models/piece.model";
import { createGame } from "src/services/game/post/post.service";
import { BadRequestError, InvalidDataError } from "src/utils/errors";
import { validateMove } from "../move.validations";

describe("Move Validations", () => {
  const pieceTypes = Object.values(PIECE_TYPE);

  pieceTypes.forEach((piece) => {
    if (piece === PIECE_TYPE.PAWN || piece === PIECE_TYPE.EMPTY) return;
    it(`should throw an Invalid Data Error when validating against a type of (${piece})`, () => {
      try {
        const moveType: MoveType = {
          from: "a1",
          to: "a2",
          color: "white",
          type: piece,
        };
        validateMove(moveType, createGame());
      } catch (err) {
        expect(err).toBeInstanceOf(InvalidDataError);
        expect(err).toMatchSnapshot({
          stack: expect.any(String),
        });
      }
    });
  });

  it("should trhow a Bad Request Error when validating against EMPTY or an unknown piece type", () => {
    try {
      const moveType: MoveType = {
        // @ts-expect-error
        type: PIECE_TYPE.EMPTY,
      };
      validateMove(moveType, createGame());
    } catch (err) {
      expect(err).toBeInstanceOf(BadRequestError);
      expect(err).toMatchInlineSnapshot(
        {
          stack: expect.any(String),
        },
        `
{
  "message": "Unsupported piece type. Expecting one of: (BISHOP,KING,KNIGHT,PAWN,QUEEN,ROOK)",
  "stack": Any<String>,
  "statusCode": 400,
}
`
      );
    }
  });

  // TODO: Fill this out!
  describe("PAWN Move Validations", () => {});
});
