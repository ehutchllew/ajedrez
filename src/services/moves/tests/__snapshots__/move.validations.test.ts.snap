// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Move Validations should throw an Invalid Data Error when validating against a type of (BISHOP) 1`] = `
{
  "message": "Bishop piece not yet supported",
  "stack": Any<String>,
  "statusCode": 422,
}
`;

exports[`Move Validations should throw an Invalid Data Error when validating against a type of (KING) 1`] = `
{
  "message": "King piece not yet supported",
  "stack": Any<String>,
  "statusCode": 422,
}
`;

exports[`Move Validations should throw an Invalid Data Error when validating against a type of (KNIGHT) 1`] = `
{
  "message": "Knight piece not yet supported",
  "stack": Any<String>,
  "statusCode": 422,
}
`;

exports[`Move Validations should throw an Invalid Data Error when validating against a type of (QUEEN) 1`] = `
{
  "message": "Queen piece not yet supported",
  "stack": Any<String>,
  "statusCode": 422,
}
`;

exports[`Move Validations should throw an Invalid Data Error when validating against a type of (ROOK) 1`] = `
{
  "message": "Rook piece not yet supported",
  "stack": Any<String>,
  "statusCode": 422,
}
`;
